core/src/main/java/org/testcontainers/images/RemoteDockerImage.java

```diff
@@ -48,58 +48,64 @@ protected final String resolve() {
             profiler.start("Check local images");
 
             int attempts = 0;
+            DockerClientException lastException = null;
             while (true) {
                 // Does our cache already know the image?
-                if (AVAILABLE_IMAGE_NAME_CACHE.contains(dockerImageName)) {
-                    logger.trace("{} is already in image name cache", dockerImageName);
+                if (AVAILABLE_IMAGE_NAME_CACHE.contains(imageName)) {
+                    logger.trace("{} is already in image name cache", imageName);
                     break;
                 }
 
                 // Update the cache
                 ListImagesCmd listImagesCmd = dockerClient.listImagesCmd();
 
                 if (Boolean.parseBoolean(System.getProperty("useFilter"))) {
-                    listImagesCmd = listImagesCmd.withImageNameFilter(dockerImageName);
+                    listImagesCmd = listImagesCmd.withImageNameFilter(imageName.toString());
                 }
 
                 List<Image> updatedImages = listImagesCmd.exec();
-                for (Image image : updatedImages) {
-                    if (image.getRepoTags() != null) {
-                        Collections.addAll(AVAILABLE_IMAGE_NAME_CACHE, image.getRepoTags());
-                    }
-                }
+                updatedImages.stream()
+                    .map(Image::getRepoTags)
+                    .filter(Objects::nonNull)
+                    .flatMap(Stream::of)
+                    .map(DockerImageName::new)
+                    .forEach(AVAILABLE_IMAGE_NAME_CACHE::add);
```

Comment:
`forEach` here is inefficient, let's use:
```java
.collect(Collectors.toCollection(() -> AVAILABLE_IMAGE_NAME_CACHE));
```

Reply:
TIL yet another stream API usage :D